Unify reproject behaviour on Dataset and DataArray

Kirill888 opened the pull request:
Expose all the possible `GeoBox.from_bbox` parameters on `.reproject`, `.output_geobox`,  `.to_crs`

- Closes #122

Kirill888 committed:
Move AnchorEnum to top-level

Kirill888 committed:
Expand functionality of compute_output_geobox

Support all the options understood by `GeoBox.from_bbox`:
anchor, shape, tol.

Kirill888 committed:
Expose output_geobox params on GeoBox.to_crs

Kirill888 committed:
Pass all args to compute_output_geobox #122

Kirill888 committed:
Fix up seealso markers in docs

Kirill888 committed:
Bump netlify action version

robbibt commented:
Still reviewing this, but a small issue I've found so far: `shape` isn't respected if you try to reproject to the same CRS but different shape:

```
import rioxarray
import odc.geo.xr

ds = rioxarray.open_rasterio(
    filename="https://data.dea.ga.gov.au/derivative/ga_ls_wo_fq_cyear_3/1-6-0/x30/y24/1995--P1Y/ga_ls_wo_fq_cyear_3_x30y24_1995--P1Y_final_frequency.tif",
    band_as_variable=True,
    chunks={},
)

ds_reprojected = ds.odc.reproject(how="EPSG:3577", shape=(100, 100))

print(ds.odc.geobox.shape)
print(ds_reprojected.odc.geobox.shape)
```
![image](https://github.com/opendatacube/odc-geo/assets/17680388/caf25a83-d142-4b58-b432-d1a1fb64121d)

Kirill888 committed:
Extend `anchor=` to take literal strings and float

center|centre|0.5 -> CENTER enum
edge|0 -> EDGE enum
floating -> FLOATING enum
float(a) -> xy_(a, a)

robbibt commented:
And a more general comment: I find it difficult to find out what params are actually supported by `.odc.reproject()`. When I try and inspect the available docstring (e.g. shift tab) I get this:
![image](https://github.com/opendatacube/odc-geo/assets/17680388/316b22f9-aa41-46e9-b702-15956a8d57c4)

...which doesn't provide any mention of the new `GeoBox.from_bbox` params that are now supported (e.g. `resolution`, `anchor`, `shape` etc). 


Is there any way we can get a more detailed/helpful docstring to appear here (or include the new params with default values in the function definition so they are at least visible)? Trying to work out what params are supported by `.odc.()` extension methods is something that other users in DEA have struggled with too - having more descriptive/easy to access docs could really help improve usability here I think.

robbibt commented on odc/geo/geobox.py:
```diff
@@ -64,12 +64,40 @@
+    AnchorEnum,
+    XY[float],
+    float,
+    Literal["center"],
```
This is fantastic - will make "anchor" even more usable than the current "align". Is the double "centre" here intended?

Kirill888 replied:
yes, both are "valid" spellings

robbibt replied:
Ah missed the spelling difference, makes sense

Kirill888 committed:
Add docs to `.odc.reproject`

- Explicitly list output_geobox parameters for user visibility
- Attach reproject on ODCExtension level
- Add docs

robbibt commented on odc/geo/_xr_interop.py:
```diff
@@ -582,13 +583,61 @@ def xr_reproject(
 ) -> XrT:
     """
     Reproject raster to different projection/resolution.
 
```
```suggestion
    :param src:
      :py:class:`~xarray.Dataset` or :py:class:`~xarray.DataArray` to reproject.

    :param how:
      How to reproject the raster. Can be a GeoBox or a CRS (e.g. CRS object or
      an "ESPG:XXXX" string/integer). If a CRS is provided, the output pixel
      grid can be customised further via ``resolution``, ``shape``, ``tight``,
      ``anchor``, ``tol``, ``round_resolution``.

    :param resampling:
      Resampling method to use when reprojecting the raster. Defaults to
      "nearest", also supports "average", "bilinear", "cubic", "cubic_spline",
      "lanczos", "mode", "gauss", "max", "min", "med", "q1", "q3".

    :param dst_nodata:
      Set a custom nodata value for the output resampled raster.
```

robbibt commented:
Have finished reviewing this - everything is working really nicely. Is there an obvious solution for the `shape` issue above? Let me know if you're planning to fix it before merging this PR, or prefer to merge this as-is and solve that later (in which case I can approve).

Kirill888 committed:
Add missing parameter docs to reproject

Co-authored-by: Robbi Bishop-Taylor <Robbi.BishopTaylor@ga.gov.au>

Kirill888 commented:
> Have finished reviewing this - everything is working really nicely. Is there an obvious solution for the `shape` issue above? Let me know if you're planning to fix it before merging this PR, or prefer to merge this as-is and solve that later (in which case I can approve).

I'd rather make sure that case has a test and works first. I can merge it without approve anyway, so no issue there.

Thanks for doc suggestions.

Kirill888 committed:
Fix in output_geobox "same geobox" logic

Check all user-supplied parameters before deciding to default
to source geobox as output.

Needed to add "default" literal to GeoboxAnchor to distinguish requested
edge from default value.

Kirill888 commented:
@robbibt this should all be fixed now, I'll merge now.
